File: travel_deals/dates.py
from __future__ import annotations

from datetime import date, timedelta

from .config import DateHorizon
from .models import DateWindow


def add_months(value: date, months: int) -> date:
    month = value.month - 1 + months
    year = value.year + month // 12
    month = month % 12 + 1
    day = min(value.day, _days_in_month(year, month))
    return date(year, month, day)


def _days_in_month(year: int, month: int) -> int:
    if month == 12:
        return 31
    return (date(year, month + 1, 1) - timedelta(days=1)).day


def generate_date_windows(
    today: date,
    horizon: DateHorizon,
    stay_lengths: tuple[int, ...],
) -> list[DateWindow]:
    start = add_months(today, horizon.start_months_from_now)
    end = add_months(today, horizon.end_months_from_now)
    current = start
    windows: list[DateWindow] = []
    while current <= end:
        for nights in stay_lengths:
            windows.append(DateWindow(current, current + timedelta(days=nights), nights))
        current += timedelta(days=horizon.step_days)
    return windows

File: src/dates.ts
import type { DateHorizon } from "./config.js";
import type { DateWindow } from "./models.js";

const MS_PER_DAY = 24 * 60 * 60 * 1000;

export function addMonths(value: Date, months: number): Date {
  const year = value.getUTCFullYear();
  const month = value.getUTCMonth();
  const day = value.getUTCDate();
  const target = new Date(Date.UTC(year, month + months, 1));
  const lastDay = daysInMonth(target.getUTCFullYear(), target.getUTCMonth());
  target.setUTCDate(Math.min(day, lastDay));
  return target;
}

export function addDays(value: Date, days: number): Date {
  return new Date(value.getTime() + days * MS_PER_DAY);
}

export function dateOnly(value: Date): string {
  return value.toISOString().slice(0, 10);
}

export function parseDateOnly(value: string): Date {
  return new Date(`${value}T00:00:00.000Z`);
}

export function generateDateWindows(today: Date, horizon: DateHorizon, stayLengths: number[]): DateWindow[] {
  const start = addMonths(today, horizon.startMonthsFromNow);
  const end = addMonths(today, horizon.endMonthsFromNow);
  const windows: DateWindow[] = [];
  let current = start;
  while (current.getTime() <= end.getTime()) {
    for (const nights of stayLengths) {
      const checkOut = addDays(current, nights);
      const checkInLabel = dateOnly(current);
      const checkOutLabel = dateOnly(checkOut);
      windows.push({
        checkIn: checkInLabel,
        checkOut: checkOutLabel,
        nights,
        label: `${checkInLabel} to ${checkOutLabel}`
      });
    }
    current = addDays(current, horizon.stepDays);
  }
  return windows;
}

function daysInMonth(year: number, zeroBasedMonth: number): number {
  return new Date(Date.UTC(year, zeroBasedMonth + 1, 0)).getUTCDate();
}
